File: projects/agentic-automation-framework-python/agentic/agent/queue.py
from concurrent.futures import ThreadPoolExecutor


class RunQueue:
    def __init__(self, worker, concurrency=2, retries=1, on_result=None):
        self.worker = worker
        self.concurrency = concurrency
        self.retries = retries
        self.on_result = on_result or (lambda result, job: None)
        self.pool = ThreadPoolExecutor(max_workers=concurrency)

    def add(self, job):
        self.pool.submit(self._run_with_retry, job)

    def _run_with_retry(self, job):
        attempt = 0
        result = None
        while attempt <= self.retries:
            result = self.worker(job)
            if result.get("ok"):
                break
            attempt += 1
        self.on_result(result, job)

File: projects/agentic-automation-framework/src/agent/queue.js
export class RunQueue {
  constructor({ worker, concurrency = 2, retries = 1, onResult = () => {} }) {
    this.worker = worker;
    this.concurrency = concurrency;
    this.retries = retries;
    this.onResult = onResult;
    this.queue = [];
    this.active = 0;
  }

  add(job) {
    this.queue.push({ ...job, attempt: 0 });
    this._drain();
  }

  _drain() {
    while (this.active < this.concurrency && this.queue.length) {
      const job = this.queue.shift();
      this._run(job);
    }
  }

  async _run(job) {
    this.active++;
    let result;
    try {
      result = await this.worker(job);
      if (!result?.ok && job.attempt < this.retries) {
        this.queue.push({ ...job, attempt: job.attempt + 1 });
      } else {
        this.onResult(result, job);
      }
    } catch (e) {
      if (job.attempt < this.retries) {
        this.queue.push({ ...job, attempt: job.attempt + 1 });
      } else {
        this.onResult({ ok: false, error: String(e), name: job.name || "job" }, job);
      }
    } finally {
      this.active--;
      this._drain();
    }
  }
}
